File: crates/blazen-py/examples/basic_workflow.py
"""Basic 3-step Blazen workflow example.

Demonstrates the core Blazen Python API by building a simple greeting
pipeline with three steps:

    StartEvent -> parse_input -> GreetEvent
    GreetEvent -> transform   -> FormattedEvent
    FormattedEvent -> greet   -> StopEvent

Each step receives a Context (shared key/value store) and an Event,
then returns a new Event to route to the next step. The workflow
finishes when a StopEvent is returned.

Run with: python basic_workflow.py
"""

import asyncio

from blazen import Context, Event, StartEvent, StopEvent, Workflow, step


# Step 1: Parse the incoming start payload.
# By default, @step accepts StartEvent, so no `accepts` argument is needed.
@step
async def parse_input(ctx: Context, ev: Event) -> Event:
    name = ev.name
    greeting_style = ev.style

    # Store the original name in shared context for later steps.
    ctx.set("original_name", name)

    return Event("GreetEvent", name=name, style=greeting_style)


# Step 2: Transform the greeting into a formatted message.
# Accepts only GreetEvent -- it will not fire on StartEvent or other types.
@step(accepts=["GreetEvent"])
async def transform(ctx: Context, ev: Event) -> Event:
    name = ev.name
    style = ev.style

    if style == "formal":
        message = f"Good day, {name}. It is a pleasure to meet you."
    elif style == "casual":
        message = f"Hey {name}, what's up!"
    else:
        message = f"Hello, {name}!"

    return Event("FormattedEvent", message=message)


# Step 3: Finalize and stop the workflow.
# Returning a StopEvent ends execution and delivers the result.
@step(accepts=["FormattedEvent"])
async def greet(ctx: Context, ev: Event) -> StopEvent:
    # Read back from shared context to show ctx.get() usage.
    original_name = ctx.get("original_name")

    return StopEvent(result={
        "greeting": ev.message,
        "original_name": original_name,
    })


async def main() -> None:
    # Build the workflow from the three steps.
    wf = Workflow("greeting-pipeline", [parse_input, transform, greet])

    # Run the workflow. Keyword arguments become the StartEvent payload.
    handler = await wf.run(name="Alice", style="formal")

    # Await the final StopEvent.
    result = await handler.result()

    print("Workflow complete!")
    print(f"  Result: {result.to_dict()}")


if __name__ == "__main__":
    asyncio.run(main())

File: crates/blazen-node/examples/basic_workflow.ts
/**
 * Basic Multi-Step Workflow
 *
 * Demonstrates a three-step Blazen workflow that processes user input
 * through a pipeline: parsing, transforming, and producing a final greeting.
 *
 * Run with: npx tsx basic_workflow.ts
 */

import { Workflow } from "blazen";
import type { Context, JsWorkflowResult } from "blazen";

const wf = new Workflow("greeting-pipeline");

wf.addStep("parse_input", ["blazen::StartEvent"], async (event: Record<string, any>, ctx: Context) => {
  const name: string = event.name || "World";
  await ctx.set("received_at", new Date().toISOString());
  return {
    type: "GreetEvent",
    name: name,
  };
});

wf.addStep("transform", ["GreetEvent"], async (event: Record<string, any>, ctx: Context) => {
  const capitalized: string = event.name.charAt(0).toUpperCase() + event.name.slice(1);
  const greeting: string = `Hello, ${capitalized}!`;
  return {
    type: "FormattedEvent",
    greeting: greeting,
    originalName: event.name,
  };
});

wf.addStep("greet", ["FormattedEvent"], async (event: Record<string, any>, ctx: Context) => {
  const receivedAt: string = await ctx.get("received_at");
  return {
    type: "blazen::StopEvent",
    result: {
      greeting: event.greeting,
      originalName: event.originalName,
      receivedAt: receivedAt,
    },
  };
});

const result: JsWorkflowResult = await wf.run({ name: "blazen" });

console.log("Workflow finished!");
console.log("  type:", result.type);
console.log("  greeting:", result.data.greeting);
console.log("  originalName:", result.data.originalName);
console.log("  receivedAt:", result.data.receivedAt);
